trie/verkle.go

```diff
@@ -0,0 +1,384 @@
+	case *verkle.InternalNode:
+		return root.InsertStem(key, values, trie.FlatdbNodeResolver)
+	default:
+		panic("invalid root type")
+	}
+}
+
+// Update associates key with value in the trie. If value has length zero, any
+// existing value is deleted from the trie. The value bytes must not be modified
+// by the caller while they are stored in the trie. If a node was not found in the
+// database, a trie.MissingNodeError is returned.
+func (trie *VerkleTrie) UpdateStorage(address common.Address, key, value []byte) error {
+	k := utils.GetTreeKeyStorageSlotWithEvaluatedAddress(trie.pointCache.GetTreeKeyHeader(address[:]), key)
+	var v [32]byte
+	if len(value) >= 32 {
+		copy(v[:], value[:32])
+	} else {
+		copy(v[32-len(value):], value[:])
+	}
+	return trie.root.Insert(k, v[:], trie.FlatdbNodeResolver)
+}
+
+func (t *VerkleTrie) DeleteAccount(addr common.Address) error {
+	var (
+		err    error
+		values = make([][]byte, verkle.NodeWidth)
+		stem   = t.pointCache.GetTreeKeyVersionCached(addr[:])
+	)
+
+	for i := 0; i < verkle.NodeWidth; i++ {
+		values[i] = zero[:]
+	}
+
+	switch root := t.root.(type) {
+	case *verkle.InternalNode:
+		err = root.InsertStem(stem, values, t.FlatdbNodeResolver)
+	default:
+		return errInvalidRootType
+	}
+	if err != nil {
+		return fmt.Errorf("DeleteAccount (%x) error: %v", addr, err)
+	}
+	// TODO figure out if the code size needs to be updated, too
+
+	return nil
+}
+
+// Delete removes any existing value for key from the trie. If a node was not
+// found in the database, a trie.MissingNodeError is returned.
+func (trie *VerkleTrie) DeleteStorage(addr common.Address, key []byte) error {
+	pointEval := trie.pointCache.GetTreeKeyHeader(addr[:])
+	k := utils.GetTreeKeyStorageSlotWithEvaluatedAddress(pointEval, key)
+	var zero [32]byte
+	return trie.root.Insert(k, zero[:], trie.FlatdbNodeResolver)
+}
+
+// Hash returns the root hash of the trie. It does not write to the database and
+// can be used even if the trie doesn't have one.
+func (trie *VerkleTrie) Hash() common.Hash {
+	return trie.root.Commit().Bytes()
+}
+
+func nodeToDBKey(n verkle.VerkleNode) []byte {
+	ret := n.Commitment().Bytes()
+	return ret[:]
+}
+
+// Commit writes all nodes to the trie's memory database, tracking the internal
+// and external (for account tries) references.
+func (trie *VerkleTrie) Commit(_ bool) (common.Hash, *trienode.NodeSet, error) {
+	root, ok := trie.root.(*verkle.InternalNode)
+	if !ok {
+		return common.Hash{}, nil, errors.New("unexpected root node type")
+	}
+	nodes, err := root.BatchSerialize()
+	if err != nil {
+		return common.Hash{}, nil, fmt.Errorf("serializing tree nodes: %s", err)
+	}
+
+	nodeset := trienode.NewNodeSet(common.Hash{})
+	for _, node := range nodes {
+		// hash parameter is not used in pathdb
+		nodeset.AddNode(node.Path, trienode.New(common.Hash{}, node.SerializedBytes))
+	}
+
+	// Serialize root commitment form
+	trie.rootHash = trie.Hash()
+	return trie.rootHash, nodeset, nil
```

Comment:
@rjl493456442 note that I now use the proper hash, and also save it so that it can be used in `OpenTrie` to load the tree from the DB.